helpers/subscriptionControllers.js

```diff
@@ -0,0 +1,140 @@
+import { cssEscape } from '../../helpers/dom.js';
+
+export class ProviderController {
+
+	constructor(host, callbacks, options) {
+		this._host = host;
+		this._callbacks = callbacks;
+		this._eventName = options.eventName;
+		this.subscribers = new Map();
+
+		this._handleSubscribe = this._handleSubscribe.bind(this);
+	}
+
+	hostConnected() {
+		if (this._eventName) this._host.addEventListener(this._eventName, this._handleSubscribe);
+	}
+
+	hostDisconnected() {
+		if (this._eventName) this._host.removeEventListener(this._eventName, this._handleSubscribe);
+	}
+
+	subscribe(target) {
+		if (this.subscribers.has(target)) return;
+		this.subscribers.set(target, target);
+		this.updateSubscribers();
+		if (this._callbacks.onSubscribe) this._callbacks.onSubscribe(target);
+	}
+
+	unsubscribe(target) {
+		this.subscribers.delete(target);
+		if (this._callbacks.onUnsubscribe) this._callbacks.onUnsubscribe(target);
+	}
+
+	updateSubscribers() {
+		if (!this.subscribers || this.subscribers.size === 0) return;
+		if (!this._callbacks.updateSubscribers) return;
+
+		// debounce the updates
+		if (this._updateSubscribersRequested) return;
+
+		this._updateSubscribersRequested = true;
+		setTimeout(() => {
+			this._callbacks.updateSubscribers(this.subscribers);
+			this._updateSubscribersRequested = false;
+		}, 0);
+	}
+
+	_handleSubscribe(e) {
+		e.stopPropagation();
+		e.detail.provider = this._host;
+		const target = e.composedPath()[0];
+		this.subscribe(target);
+	}
+}
+
+export class SubscriberController {
```

Comment:
I don't think this contreoller reads super well on how you can use _either_ the event method or the `for` method because the consumer of the controller can set themselves up for both.  Tried to add some comments throughout

Reply:
Yeah they're fairly different.  The tricky part is that selection components can use either, depending on whether the `for` property is set, and we don't know which method is being used in the constructor.  But attaching both controllers and then just using one of them sounds cleaner, it's probably not a common use case.  Filter would just use the `ForPropertySubscriberController`

Reply:
Cool I like that.